Remap paths in UI tests by default

TimNN opened the pull request:
If you think this needs further discussions / something RFC-like, please let me know the best forum for that.

This PR runs UI tests with a remapped "src base" directory by default.

Why? Because some UI tests currently depend on the length of the absolute path to the `src/test/ui` directory. Remapping makes the tests independent of the absolute path. 

The path to the source file (which is absolute on CI) is part of the type name of closures. `rustc` diagnostic output depends on the length of type names (long type names are truncated). So a long absolute path leads to long closure type names, which leads to truncation and changed diagnostics.

(I initially tried just disabling type name truncation, but that made some error messages stupid long (thousands of characters, IIRC)).

Additional changes:

* All boolean `compiletest` directives now support explicit `no-` versions to disable them.
* Adapt existing tests when necessary:
  * Disable remapping for individual tests that fail with it enabled (when there's no obvious alternative fix).
  * For tests that already check something remapping related switch to the new option unless we gain something significant by keeping the manual remap.

Passed Windows CI in https://github.com/rust-lang/rust/actions/runs/3933100590

ChrisDenton commented:
> (edit: Though I see now that Windows seems to only run for auto so I guess I'll have to figure something else out for that).

See the [Using CI to test](https://rustc-dev-guide.rust-lang.org/tests/ci.html#using-ci-to-test) section of the dev guide for how to test Windows jobs in a PR.

TimNN commented:
@ChrisDenton: Thanks for pointing out the docs! I though something like that would be possible, but didn't realize it was officially documented. (I _just_ got back to this PR and finished setting up a windows VM...)

TimNN committed:
compiletest: Generic support for "no-" versions of boolean directives.

TimNN committed:
compiletest: Support running with a remapped base directory.

TimNN committed:
ui tests: Remap test base directory by default.

TimNN commented:
r? bootstrap

jyn514 commented:
r? @Mark-Simulacrum

TimNN commented:
I've been considering this a bit more and will be moving the first two commits to separate PRs (rental & path mapping undo).

Mark-Simulacrum commented:
@rustbot author

bors commented:
:umbrella: The latest upstream changes (presumably #106853) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

TimNN commented:
* Rebased on latest master, on top of the split-out rental and path mapping changes.
* Updated more test cases to use the new option instead of manual remapping.
* Updated the PR description in the first comment.
* Temporarily added the Windows CI again.

Should generally be ready for review, I'll ping the bot to updated the labels once Windows CI is green.

TimNN commented:
@rustbot ready

Mark-Simulacrum commented:
> So a long absolute path leads to long closure type names, which leads to truncation and changed diagnostics.

Hm, I thought that we had a separate filtering pass which made this not true - https://github.com/rust-lang/rust/pull/96551?

Or is the truncation happening within rustc?

TimNN commented:
The truncation is happening in `rustc`. It shows a very abbreviated type name and includes another note stating something along the lines of "the full type name has been written to <some tmp path>".

TimNN commented:
Here is an example diff we see if the filename is too long:

```
---- [ui] rust-llvm-integrate-prototype/tests/ui/never_type/fallback-closure-wrap.rs#fallback stdout ----

diff of stderr:

10	   |
11	   = note: expected unit type `()`
12	                   found type `!`
-	   = note: required for the cast from `[closure@$DIR/fallback-closure-wrap.rs:18:40: 18:47]` to the object type `dyn FnMut()`
+	   = note: required for the cast from `[closure@fallback-closure-wrap.rs:18:40]` to the object type `dyn FnMut()`
+	   = note: the full name for the casted type has been written to '$TEST_BUILD_DIR/never_type/fallback-closure-wrap.fallback/fallback-closure-wrap.long-type-6621918892693463368.txt'
14	
15	error: aborting due to previous error
16	
```

Mark-Simulacrum commented:
Huh! This is news to me, maybe a recent change.

In any case, this seems like a reasonable approach to solving that problem -- @bors r+ rollup=iffy

bors commented:
:pushpin: Commit cd1d0bc20cd2c18ec844f6447396dadf90b762e9 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum cd1d0bc20cd2c18ec844f6447396dadf90b762e9 -->
<!-- homu: {"type":"Approved","sha":"cd1d0bc20cd2c18ec844f6447396dadf90b762e9","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit cd1d0bc20cd2c18ec844f6447396dadf90b762e9 with merge 52372f9c71d8ade4cb815524f179119656f0aa2e...
<!-- homu: {"type":"BuildStarted","head_sha":"cd1d0bc20cd2c18ec844f6447396dadf90b762e9","merge_sha":"52372f9c71d8ade4cb815524f179119656f0aa2e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3975903282/jobs/6817118799)
Approved by: Mark-Simulacrum
Pushing 52372f9c71d8ade4cb815524f179119656f0aa2e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3975903282/jobs/6817118799"},"merge_sha":"52372f9c71d8ade4cb815524f179119656f0aa2e"} -->

rust-timer commented:
Finished benchmarking commit (52372f9c71d8ade4cb815524f179119656f0aa2e): [comparison URL](https://perf.rust-lang.org/compare.html?start=005fc0f00f2d4ceaf523b67a8f9c5665b8ac5baf&end=52372f9c71d8ade4cb815524f179119656f0aa2e&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=005fc0f00f2d4ceaf523b67a8f9c5665b8ac5baf&end=52372f9c71d8ade4cb815524f179119656f0aa2e&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.3%, -0.3%] | 2     |
| Improvements ✅ <br /> (secondary) | -0.6% | [-0.8%, -0.3%] | 7     |
| All ❌✅ (primary)                 | -0.3% | [-0.3%, -0.3%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=005fc0f00f2d4ceaf523b67a8f9c5665b8ac5baf&end=52372f9c71d8ade4cb815524f179119656f0aa2e&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.1% | [-5.1%, -5.1%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=005fc0f00f2d4ceaf523b67a8f9c5665b8ac5baf&end=52372f9c71d8ade4cb815524f179119656f0aa2e&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
